pkg/gardenlet/controller/backupentry/backup_entry_reconciler.go

```diff
@@ -110,6 +110,11 @@ func (r *reconciler) reconcileBackupEntry(ctx context.Context, gardenClient kube
 		operationType = gardencorev1beta1.LastOperationTypeRestore
 	}
 
+	if updateErr := removeGardenerOperationAnnotation(ctx, gardenClient.Client(), backupEntry); updateErr != nil {
+		backupEntryLogger.Errorf("Could not remove %q annotation: %+v", v1beta1constants.GardenerOperation, updateErr)
+		return reconcile.Result{}, updateErr
+	}
```

Comment:
Let's say the object was annotated with the `restore` operation and the `gardenlet` exits right after those lines. When it comes up again, wouldn't it start performing a different operation because the information that it should actually be a "restore" is lost?

Reply:
Yes, unfortunately it would. I guess the best way to deal with it is to update the status and remove the annotation in the same `Patch` operation. Is that ever possible, or do you have other suggestions how to handle it?

I also thought about the following sequence:
* Determine `operationType` based on the operation annotation.
* Update `status.lastOperation` (but not `status.observedGeneration`) yet.
* Remove the operation annotation.
* Update `status.observedGeneration`.
But I think this will enqueue the BackupEntry after the first status update. Still, perhaps it's better than what we have now.

Let me play a bit more with the above ideas.

Reply:
>  I guess the best way to deal with it is to update the status and remove the annotation in the same Patch operation. Is that ever possible, or do you have other suggestions how to handle it?

No, unfortunately, this isn't possible in only one call.

Can't we update the `.status.lastOperation` to `<OPERATION> Progressing` first (including updating `.status.observedGeneration`) and remove the annotation afterwards? If the gardenlet exits after the `.status.lastOperation` was updated then it could still start to reconcile the resource by determining the operation type based on `.status.lastOperation.type`, or?

Reply:
Removing the operation annotation after `status.observedGeneration` has been updated would lead to the original issue that I am trying to fix.

As discussed, I will look deeper into how this is handled for extension resources in the seed and perhaps try to introduce a similar handling here.

Reply:
So basically, this is what happens for extension resources in the seed:
* Instead of checking the generation, resources are enqueued if they have the operation annotation, if the last operation was not successful, or if the deletion timestamp is not nil.
* The operation annotation is removed before the reconciliation if it's `reconcile`, and at the end of the reconciliation if it's `migrate` or `restore`.

This can lead to the following issues:
* If the extension is restarted after removing the operation annotation with value `reconcile`, but before updating the status, the reconciliation will be skipped. This is perhaps ok though since gardenlet will retry the flow at some point leading to another reconciliation.
* If the extension is restarted after updating the status, but before removing the operation annotation with value `restore`, or `migrate`, one redundant reconciliation will be enqueued. I guess this is also ok, since it's rare, and it would only happen for `restore` and `migrate` (the handling should be idempotent by design).

However, it cannot lead to either the original issue reported in #4540 (since the annotation itself, and not the generation, is what matters), or the issue you with my current fix (since the annotation is not removed if it's value is `restore` or `migrate`).

So basically what I would propose to do is to rework the enqueueing of `BackupEntry` to follow the above pattern, and even to get rid of checking the operation annotation in `strategy.go` (since the generation is no longer used to determine whether the resource should be enqueued).

Reply:
Sounds good to me, thanks for looking into it!

Reply:
OK, did as proposed above, please take another look.